masterslave.q

```diff
@@ -1,62 +1,88 @@
+// should then become master
+//3 processes, one master, master dies, one of the remaining two takes over
+
+init:{
+ // the list of processes to connect to
+ .servers.CONNECTIONS:distinct .servers.CONNECTIONS,.proc.proctype;
+
+ // custom function, this is invoked when a new outbound connection is created
+ // to be customised to invoke negotiation of processes
+ .servers.connectcustom:{.lg.o[`connect;"found new connection"];
+                         .masterslave.checkifmaster[];
+                         };
+ // create connections
+ .servers.startup[]
+ };
 
 \d .masterslave
 
 //table scehma of connected lb processes with ismaster status
-statustable:([handle:`int$()] procname:`symbol$();starttimeUTC:`timestamp$();ismaster:`boolean$())
+statustable:([handle:`int$()] procname:();starttimeUTC:`timestamp$();ismaster:`boolean$());
 
 //store own process start timestamp
-start:.z.p
+start:.z.p;
 
 //set details dict with own details
-details:{`procname`starttimeUTC`ismaster!(.proc.procname;.masterslave.start;1b)} 
+details:{
+ `procname`starttimeUTC!(string .proc.procname;.masterslave.start)
+ };
 
 //get details of procname provided
-getdetails:{[processname] (first exec w from .servers.SERVERS where procname like processname) ".masterslave.details[]"}
+getdetails:{[processname]
+ (first exec w from .servers.SERVERS where procname like processname,not null w) (`.masterslave.details;[])
+ };
 
-//update .masterslave.statustable with other proc details and update ismaster col to determine which process is master 
+//update .masterslave.statustable with other proc details and update ismaster col to determine which process is master
 addmember:{[processname]
-    update ismaster:0b from (`.masterslave.statustable upsert .masterslave.getdetails[processname],(enlist `handle)!(enlist first exec w from .servers.SERVERS  where procname like processname)) where starttimeUTC<>min starttimeUTC}
+ `.masterslave.statustable upsert .masterslave.getdetails[processname],
+  (`handle`ismaster)!(first exec w from .servers.SERVERS  where procname like processname,not null w;1b)
+ };
+
+masterupdate:{update ismaster:starttimeUTC=min starttimeUTC from `.masterslave.statustable}
 
 //is the process itself the master
-ammaster:{exec ismaster from .masterslave.statustable where handle=0}
+ammaster:{exec ismaster from .masterslave.statustable where handle=0};
 
-//find which process is the master
-findmaster:{first exec handle from .masterslave.statustable where ismaster=1b}
+//find which process is the master - can only be run after checkifmaster has been run
+findmaster:{exec handle, procname from .masterslave.statustable where ismaster=1b};
 
 //wrapper func for finding master
 checkifmaster:{
-    .servers.startup[];
-    .masterslave.addmember each string exec procname from .servers.SERVERS where proctype like "lb", not null w
- }
+ (.masterslave.addmember')[string exec procname from .servers.SERVERS where proctype like "lb", not null w];
```

Comment:
Don't hardcode the proctype in here. Should just be 

where proctype=.proc.proctype
.i.e proctype is the same as the local process